src/Agent.Plugins/PipelineArtifact/PipelineArtifactPluginV1.cs

```diff
@@ -55,19 +56,127 @@ public class DownloadPipelineArtifactTaskV1 : PipelineArtifactTaskPluginBaseV1
+            {
+                // Project ID
+                // TODO: use a constant for project id, which is currently defined in Microsoft.VisualStudio.Services.Agent.Constants.Variables.System.TeamProjectId (Ting)
+                guidStr = context.Variables.GetValueOrDefault("system.teamProjectId")?.Value;
+                Guid.TryParse(guidStr, out Guid projectId);
+                ArgUtil.NotEmpty(projectId, nameof(projectId));
 
-            // TODO: call download given build information
+                int buildIdInt = buildId != ""  ?  Int32.Parse(buildId) : 0;
+                //Get the build ID.
+                if (int.TryParse(buildIdStr, out buildIdInt) && buildIdInt != 0)
+                {
+                    context.Output(StringUtil.Loc("DownloadingFromBuild", buildIdInt));
+                }
+                else
+                {
+                    string hostType = context.Variables.GetValueOrDefault("system.hosttype")?.Value;
+                    if (string.Equals(hostType, "Release", StringComparison.OrdinalIgnoreCase) ||
+                        string.Equals(hostType, "DeploymentGroup", StringComparison.OrdinalIgnoreCase))
+                    {
+                        throw new InvalidOperationException(StringUtil.Loc("BuildIdIsNotAvailable", hostType ?? string.Empty));
+                    }
+                    else if (!string.Equals(hostType, "Build", StringComparison.OrdinalIgnoreCase))
+                    {
+                        throw new InvalidOperationException(StringUtil.Loc("CannotDownloadFromCurrentEnvironment", hostType ?? string.Empty));
+                    }
+                    else
+                    {
+                        // This should not happen since the build id comes from build environment. But a user may override that so we must be careful.
+                        throw new ArgumentException(StringUtil.Loc("BuildIdIsNotValid", buildIdStr));
+                    }
+                } 
+                //Create the directory if absent.
+                string fullPath = Path.GetFullPath(downloadPath);
+                bool isDir = Directory.Exists(fullPath);
+                if (!isDir)
+                {
+                    Directory.CreateDirectory(fullPath);
+                }
+
+                context.Output(StringUtil.Loc("DownloadArtifactTo", downloadPath));
+                PipelineArtifactServer server = new PipelineArtifactServer();
+                await server.DownloadAsyncMinimatch(context, projectId, buildIdInt, artifactName, downloadPath, minimatchPatterns, token);
+                context.Output(StringUtil.Loc("DownloadArtifactFinished"));
+            }
+            else if(buildType == "specific")
+            {
+                // Project Name - coming from the GUI.
+                int buildIdInt = 0;
+                if(buildVersionToDownload == "latest")
+                {
+                    List<int> defns= new List<int>();
```

Comment:
this code is shared in latestFromBranch, please place it into a function

Reply:
Resolved